src/profile/hooks/useProfileMutations.ts

```diff
@@ -0,0 +1,145 @@
+import { ProfileContext, ProfileContextData } from '../context/ProfileContext';
+
+const UPDATE_PROFILE = loader('../../profile/graphql/UpdateMyProfile.graphql');
+const MY_PROFILE = loader('../../profile/graphql/MyProfileQuery.graphql');
+
+export type QueryResult = FetchResult<ProfileRoot, AnyObject, AnyObject>;
+export type UpdateResult = FetchResult<UpdateProfileRoot>;
+
+type UpdateProfileFunction = (
+  formValues: Partial<FormValues>,
+  profile?: ProfileRoot
+) => Promise<UpdateResult>;
+
+export type MutationReturnType = {
+  error: ApolloError | undefined;
+  loading: boolean;
+  profileData?: ProfileRoot;
+  update: UpdateProfileFunction;
+};
+
+type UpdatableProfileDataKeys = keyof Omit<
+  UpdateProfileData,
+  '__typename' | 'id'
+>;
+
+export type ValuesOfUpdateProfileData = UpdateProfileData[UpdatableProfileDataKeys];
+
+type CacheUpdateFunction = (
+  data?: UpdateProfileData
+) => ValuesOfUpdateProfileData;
+
+type CacheUpdateProps = Partial<
+  Record<UpdatableProfileDataKeys, CacheUpdateFunction>
+>;
+
+// cache is updated with cache.modify({fields...})
+// where properties of the "fields" object are fields to update
+// the format is: {"property": <function returning new value>}
+// example: fields:{ phones:()=>listOfPhones, primaryPhone:()=>newPrimaryPhone}
+// this helper function creates those objects from listed "fields"
+// and returns same field from given data source "updatedProfile"
+function updateFieldCreator(
+  fields: UpdatableProfileDataKeys[],
+  updatedProfile: UpdateProfileData
+): CacheUpdateProps {
+  return fields.reduce((currentValue, nextField) => {
+    currentValue[nextField] = () => updatedProfile[nextField];
+    return currentValue;
+  }, {} as CacheUpdateProps);
+}
+
+function getCacheUpdateFields(
+  dataType: EditDataType,
+  updatedProfile: UpdateProfileData
+): CacheUpdateProps {
+  const fields: UpdatableProfileDataKeys[] = [];
+  if (dataType === 'phones') {
+    fields.push(dataType, 'primaryPhone');
+  } else if (dataType === 'emails') {
+    fields.push(dataType, 'primaryEmail');
+  } else if (dataType === 'addresses') {
+    fields.push(dataType, 'primaryAddress');
+  } else if (dataType === basicDataType) {
+    fields.push('firstName', 'nickname', 'lastName');
+  } else if (dataType === additionalInformationType) {
+    fields.push('language');
+  }
+  return updateFieldCreator(fields, updatedProfile);
+}
+
+function updateCache(
+  cache: ApolloCache<UpdateProfileRoot>,
+  result: UpdateResult,
+  dataType: EditDataType,
+  profileData: ProfileRoot | null | undefined,
+  updateProfileData: ProfileContextData['updateProfileData']
+) {
+  const updatedProfile: UpdateProfileData | null | undefined =
+    result.data?.updateMyProfile?.profile;
+  const identity =
+    profileData &&
+    cache.identify((profileData.myProfile as unknown) as Reference);
+  if (!updatedProfile || !identity) {
+    throw new Error('Unable to update cache2');
```

Comment:
cache2? This might be that Im beginner related to graphql, but is there cache1 somewhere?

Reply:
That's a relic from some debugging session. :)